lib/web_ui/lib/src/engine/text/line_breaker.dart

```diff
@@ -117,30 +169,309 @@ LineBreakResult nextLineBreak(String text, int index) {
+      if (prev == LineCharProperty.SP) {
+        // LB10: Treat any remaining combining mark or ZWJ as AL.
+        curr = LineCharProperty.AL;
+      } else {
+        // Preserve the property of the previous character to treat the sequence
+        // as if it were X.
+        curr = prev;
+        continue;
+      }
+    }
+
+    // Do not break after a zero width joiner.
+    // LB8a: ZWJ ×
+    if (isPrevZWJ) {
+      continue;
+    }
+
+    // Do not break before or after Word joiner and related characters.
+    // LB11: × WJ
+    //       WJ ×
+    if (curr == LineCharProperty.WJ || prev == LineCharProperty.WJ) {
+      continue;
+    }
+
+    // Do not break after NBSP and related characters.
+    // LB12: GL ×
+    if (prev == LineCharProperty.GL) {
+      continue;
+    }
+
+    // Do not break before NBSP and related characters, except after spaces and
+    // hyphens.
+    // LB12a: [^SP BA HY] × GL
+    if (!(prev == LineCharProperty.SP ||
+            prev == LineCharProperty.BA ||
+            prev == LineCharProperty.HY) &&
+        curr == LineCharProperty.GL) {
+      continue;
+    }
+
+    // Do not break before ‘]’ or ‘!’ or ‘;’ or ‘/’, even after spaces.
+    // LB13: × CL
+    //       × CP
+    //       × EX
+    //       × IS
+    //       × SY
+    if (curr == LineCharProperty.CL ||
+        curr == LineCharProperty.CP ||
+        curr == LineCharProperty.EX ||
+        curr == LineCharProperty.IS ||
+        curr == LineCharProperty.SY) {
+      continue;
+    }
+
+    // Do not break after ‘[’, even after spaces.
+    // LB14: OP SP* ×
+    if (prev == LineCharProperty.OP ||
+        baseOfSpaceSequence == LineCharProperty.OP) {
+      continue;
+    }
+
+    // Do not break within ‘”[’, even with intervening spaces.
+    // LB15: QU SP* × OP
+    if ((prev == LineCharProperty.QU ||
+            baseOfSpaceSequence == LineCharProperty.QU) &&
+        curr == LineCharProperty.OP) {
+      continue;
+    }
+
+    // Do not break between closing punctuation and a nonstarter, even with
+    // intervening spaces.
+    // LB16: (CL | CP) SP* × NS
+    if ((prev == LineCharProperty.CL ||
+            baseOfSpaceSequence == LineCharProperty.CL ||
+            prev == LineCharProperty.CP ||
+            baseOfSpaceSequence == LineCharProperty.CP) &&
+        curr == LineCharProperty.NS) {
+      continue;
+    }
+
+    // Do not break within ‘——’, even with intervening spaces.
+    // LB17: B2 SP* × B2
+    if ((prev == LineCharProperty.B2 ||
+            baseOfSpaceSequence == LineCharProperty.B2) &&
+        curr == LineCharProperty.B2) {
+      continue;
+    }
+
+    // Break after spaces.
+    // LB18: SP ÷
+    if (prev == LineCharProperty.SP) {
+      return LineBreakResult(index, LineBreakType.opportunity);
+    }
+
+    // Do not break before or after quotation marks, such as ‘”’.
+    // LB19: × QU
+    //       QU ×
+    if (prev == LineCharProperty.QU || curr == LineCharProperty.QU) {
+      continue;
```

Comment:
Have you checked which `continue`s trigger most often in typical text? It might be worth moving those up so they short-circuit faster.

Reply:
The most common ones are probably the SP-related ones at the top, and AL x AL near the end. I've been mostly focusing on correctness and passing all the tests. I have another PR that's coming after this one that brings the last batch of tests, after that I'll be able to do perf improvements with more confidence.

Reply:
👍